--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -471,6 +_,13 @@
         return this.viewDistanceHolder;
     }
     // Paper end - rewrite chunk system
+    // SparklyPaper start - per-player chunk rate overrides
+    public @Nullable Double chunkLoadRateOverride = null;
+    public @Nullable Double chunkGenRateOverride = null;
+    public @Nullable Double chunkSendRateOverride = null;
+    public @Nullable Long maxConcurrentChunkLoadsOverride = null;
+    public @Nullable Long maxConcurrentChunkGeneratesOverride = null;
+    // SparklyPaper end
 
     public ServerPlayer(final MinecraftServer server, final ServerLevel level, final GameProfile gameProfile, final ClientInformation clientInformation) {
         super(level, gameProfile);
