--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -202,6 +_,7 @@
 
                 if (this.entity.needsSync || this.trackDelta || this.entity instanceof LivingEntity livingEntity && livingEntity.isFallFlying()) {
                     Vec3 movement = this.entity.getDeltaMovement();
+                    if (movement != this.lastSentMovement) { // SparklyPaper start - skip distanceToSqr call in ServerEntity#sendChanges if the delta movement hasn't changed
                     double diff = movement.distanceToSqr(this.lastSentMovement);
                     if (diff > 1.0E-7 || diff > 0.0 && movement.lengthSqr() == 0.0) {
                         this.lastSentMovement = movement;
@@ -219,6 +_,7 @@
                             this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
                         }
                     }
+                    } // SparklyPaper end
                 }
 
                 if (packet != null) {
