--- a/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java
+++ b/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java
@@ -85,7 +_,7 @@
     void testTextColor() {
         final TextColor color = color(0x1d38df);
         final Tag result = TEXT_COLOR_CODEC.encodeStart(NbtOps.INSTANCE, color).result().orElseThrow();
-        assertEquals("\"" + color.asHexString() + "\"", result.toString());
+        assertEquals("\"" + color.asHexString().toUpperCase(java.util.Locale.ROOT) + "\"", result.toString());
         final net.minecraft.network.chat.TextColor nms = net.minecraft.network.chat.TextColor.CODEC.decode(NbtOps.INSTANCE, result).result().orElseThrow().getFirst();
         assertEquals(color.value(), nms.getValue());
     }
