--- a/src/main/java/com/destroystokyo/paper/MaterialTags.java
+++ b/src/main/java/com/destroystokyo/paper/MaterialTags.java
@@ -83,6 +_,7 @@
      */
     public static final MaterialSetTag CONCRETES = new MaterialSetTag(keyFor("concretes"))
         .endsWith("_CONCRETE")
+        .notStartsWith("SPARKLYPOWER_") // SparklyPaper - custom blocks
         .ensureSize("CONCRETES", 16).lock();
 
     /**
@@ -152,6 +_,7 @@
      */
     public static final MaterialSetTag GLAZED_TERRACOTTA = new MaterialSetTag(keyFor("glazed_terracotta"))
         .endsWith("GLAZED_TERRACOTTA")
+        .notStartsWith("SPARKLYPOWER_") // SparklyPaper - custom blocks
         .ensureSize("GLAZED_TERRACOTTA", 16).lock();
 
     /**
