--- a/src/main/java/com/destroystokyo/paper/MaterialTags.java
+++ b/src/main/java/com/destroystokyo/paper/MaterialTags.java
@@ -183,6 +_,7 @@
         .endsWith("TERRACOTTA")
         .not(Material.TERRACOTTA)
         .notEndsWith("GLAZED_TERRACOTTA")
+        .notStartsWith("SPARKLYPOWER_") // SparklyPaper - custom blocks
         .ensureSize("STAINED_TERRACOTTA", 16).lock();
 
     /**
@@ -190,6 +_,7 @@
      */
     public static final MaterialSetTag TERRACOTTA = new MaterialSetTag(keyFor("terracotta"))
         .endsWith("TERRACOTTA")
+        .notStartsWith("SPARKLYPOWER_") // SparklyPaper - custom blocks
         .ensureSize("TERRACOTTA", 33).lock();
 
     /**
