--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1821,7 +_,15 @@
             profiler.push("tick");
 
             try {
+                long i = Util.getNanos(); // SparklyPaper - track world's MSPT
                 level.tick(haveTime);
+                // SparklyPaper start - track world's MSPT
+                long j = Util.getNanos() - i;
+                // These are from the "tickServer" function
+                level.tickTimes5s.add(this.tickCount, j);
+                level.tickTimes10s.add(this.tickCount, j);
+                level.tickTimes1m.add(this.tickCount, j);
+                // SparklyPaper end
             } catch (Throwable t) {
                 CrashReport report = CrashReport.forThrowable(t, "Exception ticking world");
                 level.fillReportDetails(report);
