--- a/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
@@ -20,7 +_,7 @@
     );
     private final int id;
     private final UUID uuid;
-    private final EntityType<?> type;
+    public EntityType<?> type; // SparklyPaper - Helpful NMS packet changes: remove final and make public
     private final double x;
     private final double y;
     private final double z;
@@ -171,6 +_,20 @@
     public float getYHeadRot() {
         return Mth.unpackDegrees(this.yHeadRot);
     }
+
+    // SparklyPaper - Helpful NMS packet changes: expose raw rotational fields
+    public byte getXRotRaw() {
+        return this.xRot;
+    }
+
+    public byte getYRotRaw() {
+        return this.yRot;
+    }
+
+    public byte getYHeadRotRaw() {
+        return this.yHeadRot;
+    }
+    // SparklyPaper end
 
     public int getData() {
         return this.data;
