--- a/net/minecraft/network/protocol/game/ClientboundBlockUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundBlockUpdatePacket.java
@@ -19,7 +_,7 @@
         ClientboundBlockUpdatePacket::new
     );
     private final BlockPos pos;
-    private final BlockState blockState;
+    public BlockState blockState; // SparklyPaper - Helpful NMS packet changes: remove final
 
     public ClientboundBlockUpdatePacket(final BlockPos pos, final BlockState state) {
         this.pos = pos;
