--- a/net/minecraft/network/protocol/game/ClientboundRotateHeadPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundRotateHeadPacket.java
@@ -21,6 +_,13 @@
         this.yHeadRot = yHeadRot;
     }
 
+    // SparklyPaper start - Helpful NMS packet changes: add entity ID constructor
+    public ClientboundRotateHeadPacket(int entityId, byte headYaw) {
+        this.entityId = entityId;
+        this.yHeadRot = headYaw;
+    }
+    // SparklyPaper end
+
     private ClientboundRotateHeadPacket(final FriendlyByteBuf input) {
         this.entityId = input.readVarInt();
         this.yHeadRot = input.readByte();
