--- a/net/minecraft/world/entity/monster/Monster.java
+++ b/net/minecraft/world/entity/monster/Monster.java
@@ -145,6 +_,12 @@
     @Override
     public ItemStack getProjectile(final ItemStack heldWeapon) {
         if (heldWeapon.getItem() instanceof ProjectileWeaponItem) {
+            // SparklyPaper start - Add EntityGetProjectileForWeaponEvent
+            net.sparklypower.sparklypaper.event.entity.EntityGetProjectileForWeaponEvent event = new net.sparklypower.sparklypaper.event.entity.EntityGetProjectileForWeaponEvent(this.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(heldWeapon));
+            event.callEvent();
+            if (event.getArrow() != null)
+                return org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getArrow());
+            // SparklyPaper end
             Predicate<ItemStack> supportedProjectiles = ((ProjectileWeaponItem)heldWeapon.getItem()).getSupportedHeldProjectiles();
             ItemStack heldProjectile = ProjectileWeaponItem.getHeldProjectile(this, supportedProjectiles);
             return heldProjectile.isEmpty() ? new ItemStack(Items.ARROW) : heldProjectile;
