--- a/src/test/java/org/bukkit/craftbukkit/legacy/LegacyTest.java
+++ b/src/test/java/org/bukkit/craftbukkit/legacy/LegacyTest.java
@@ -163,6 +_,9 @@
             Material.MUSIC_DISC_BOUNCE, Material.POLISHED_CINNABAR, Material.POLISHED_CINNABAR_SLAB, Material.POLISHED_CINNABAR_STAIRS, Material.POLISHED_CINNABAR_WALL, Material.POLISHED_SULFUR, Material.POLISHED_SULFUR_SLAB, Material.POLISHED_SULFUR_STAIRS, Material.POLISHED_SULFUR_WALL,
             Material.POTENT_SULFUR, Material.SULFUR, Material.SULFUR_BRICKS, Material.SULFUR_BRICK_SLAB, Material.SULFUR_BRICK_STAIRS, Material.SULFUR_BRICK_WALL, Material.SULFUR_CUBE_BUCKET, Material.SULFUR_CUBE_SPAWN_EGG, Material.SULFUR_SLAB, Material.SULFUR_SPIKE, Material.SULFUR_STAIRS,
             Material.SULFUR_WALL,
+            // SparklyPower custom blocks
+            Material.SPARKLYPOWER_RAINBOW_WOOL, Material.SPARKLYPOWER_RAINBOW_CONCRETE, Material.SPARKLYPOWER_RAINBOW_TERRACOTTA, Material.SPARKLYPOWER_ASPHALT_PLAYER, Material.SPARKLYPOWER_ASPHALT_SERVER, Material.SPARKLYPOWER_ASPHALT_PLAYER_SLAB,
+            Material.SPARKLYPOWER_ASPHALT_SERVER_SLAB,
             //
             Material.LEGACY_AIR, Material.LEGACY_DEAD_BUSH, Material.LEGACY_BURNING_FURNACE, Material.LEGACY_WALL_SIGN, Material.LEGACY_REDSTONE_TORCH_OFF, Material.LEGACY_SKULL, Material.LEGACY_REDSTONE_COMPARATOR_ON, Material.LEGACY_WALL_BANNER, Material.LEGACY_MONSTER_EGG));
 
