--- a/src/test/java/org/bukkit/craftbukkit/legacy/LegacyTest.java
+++ b/src/test/java/org/bukkit/craftbukkit/legacy/LegacyTest.java
@@ -177,6 +_,9 @@
             Material.PURPLE_CONCRETE_SLAB, Material.PURPLE_CONCRETE_STAIRS, Material.PURPLE_WOOL_SLAB, Material.PURPLE_WOOL_STAIRS, Material.RED_CONCRETE_SLAB, Material.RED_CONCRETE_STAIRS, Material.RED_POPLAR_LEAVES, Material.RED_SHRUB, Material.RED_WOOL_SLAB, Material.RED_WOOL_STAIRS,
             Material.SHELF_MUSHROOM, Material.STRAW_BED, Material.STRIPPED_POPLAR_LOG, Material.STRIPPED_POPLAR_WOOD, Material.WHITE_CONCRETE_SLAB, Material.WHITE_CONCRETE_STAIRS, Material.WHITE_WOOL_SLAB, Material.WHITE_WOOL_STAIRS, Material.YELLOW_CONCRETE_SLAB, Material.YELLOW_CONCRETE_STAIRS,
             Material.YELLOW_POPLAR_LEAVES, Material.YELLOW_WOOL_SLAB, Material.YELLOW_WOOL_STAIRS,
+            // SparklyPower custom blocks
+            Material.SPARKLYPOWER_RAINBOW_WOOL, Material.SPARKLYPOWER_RAINBOW_CONCRETE, Material.SPARKLYPOWER_RAINBOW_TERRACOTTA, Material.SPARKLYPOWER_ASPHALT_PLAYER, Material.SPARKLYPOWER_ASPHALT_SERVER, Material.SPARKLYPOWER_ASPHALT_PLAYER_SLAB,
+            Material.SPARKLYPOWER_ASPHALT_SERVER_SLAB,
             //
             Material.LEGACY_AIR, Material.LEGACY_DEAD_BUSH, Material.LEGACY_BURNING_FURNACE, Material.LEGACY_WALL_SIGN, Material.LEGACY_REDSTONE_TORCH_OFF, Material.LEGACY_SKULL, Material.LEGACY_REDSTONE_COMPARATOR_ON, Material.LEGACY_WALL_BANNER, Material.LEGACY_MONSTER_EGG));
 
