--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -5846,6 +_,52 @@
             .sound(SoundType.SWEET_BERRY_BUSH)
             .pushReaction(PushReaction.DESTROY)
     );
+    // SparklyPaper start - SparklyPower Survival custom blocks
+    // Blocks' strength should be synced with the replaced block on the client side!
+    public static final Block SPARKLYPOWER_RAINBOW_WOOL = register(
+        BlockItemIds.SPARKLYPOWER_RAINBOW_WOOL,
+        Block::new,
+        BlockBehaviour.Properties.of().mapColor(MapColor.TERRACOTTA_RED).strength(0.5F).sound(SoundType.GRASS)
+    );
+    public static final Block SPARKLYPOWER_RAINBOW_CONCRETE = register(
+        BlockItemIds.SPARKLYPOWER_RAINBOW_CONCRETE,
+        Block::new,
+        BlockBehaviour.Properties.of().mapColor(MapColor.TERRACOTTA_RED).strength(0.5F).sound(SoundType.GRASS)
+    );
+    public static final Block SPARKLYPOWER_RAINBOW_TERRACOTTA = register(
+        BlockItemIds.SPARKLYPOWER_RAINBOW_TERRACOTTA,
+        Block::new,
+        BlockBehaviour.Properties.of().mapColor(MapColor.TERRACOTTA_RED).strength(0.5F).sound(SoundType.GRASS)
+    );
+    public static final Block SPARKLYPOWER_ASPHALT_SERVER = register(
+        BlockItemIds.SPARKLYPOWER_ASPHALT_SERVER,
+        Block::new,
+        BlockBehaviour.Properties.of()
+            .mapColor(MapColor.COLOR_BLACK)
+            .instrument(NoteBlockInstrument.BASEDRUM)
+            .requiresCorrectToolForDrops()
+            .strength(1.5F, 6.0F)
+    );
+    public static final Block SPARKLYPOWER_ASPHALT_SERVER_SLAB = register(
+        BlockItemIds.SPARKLYPOWER_ASPHALT_SERVER_SLAB,
+        SlabBlock::new,
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_BLACK).instrument(NoteBlockInstrument.BASEDRUM).requiresCorrectToolForDrops().strength(2.0F, 6.0F)
+    );
+    public static final Block SPARKLYPOWER_ASPHALT_PLAYER = register(
+        BlockItemIds.SPARKLYPOWER_ASPHALT_PLAYER,
+        Block::new,
+        BlockBehaviour.Properties.of()
+            .mapColor(MapColor.COLOR_BLACK)
+            .instrument(NoteBlockInstrument.BASEDRUM)
+            .requiresCorrectToolForDrops()
+            .strength(1.5F, 6.0F)
+    );
+    public static final Block SPARKLYPOWER_ASPHALT_PLAYER_SLAB = register(
+        BlockItemIds.SPARKLYPOWER_ASPHALT_PLAYER_SLAB,
+        SlabBlock::new,
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_BLACK).instrument(NoteBlockInstrument.BASEDRUM).requiresCorrectToolForDrops().strength(2.0F, 6.0F)
+    );
+    // SparklyPaper end
 
     private static ToIntFunction<BlockState> litBlockEmission(final int lightEmission) {
         return state -> state.getValue(BlockStateProperties.LIT) ? lightEmission : 0;
