--- a/net/minecraft/world/level/block/StemBlock.java
+++ b/net/minecraft/world/level/block/StemBlock.java
@@ -69,6 +_,56 @@
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (level.getRawBrightness(pos, 0) >= 9) {
+            // SparklyPaper start - Blazingly simple farm checks
+            if (level.sparklyPaperConfig.getBlazinglySimpleFarmChecks().getEnabled()) {
+                // These checks are similar to getGrowthSpeed, but we have "inlined" them because we want to access stuff like the farm block data later on
+                // Is the block below us moisturised?
+                BlockPos farmlandBelowTheCurrentBlock = pos.below();
+                BlockState farmlandBelowTheCurrentBlockData = level.getBlockState(farmlandBelowTheCurrentBlock);
+                float f = level.sparklyPaperConfig.getBlazinglySimpleFarmChecks().getDefaultGrowthSpeed();
+                boolean isCurrentFarmlandStateMoist = false;
+                if (farmlandBelowTheCurrentBlockData.is(Blocks.FARMLAND)) {
+                    if ((Integer) farmlandBelowTheCurrentBlockData.getValue(FarmBlock.MOISTURE) > 0) {
+                        // If we are currently moist, increase the speed!
+                        f = level.sparklyPaperConfig.getBlazinglySimpleFarmChecks().getMoistGrowthSpeed();
+                        isCurrentFarmlandStateMoist = true;
+                    }
+                }
+
+                if (random.nextFloat() < ((this == Blocks.PUMPKIN_STEM ? level.spigotConfig.pumpkinModifier : level.spigotConfig.melonModifier) / (100.0f * (Math.floor((25.0F / f) + 1))))) { // Spigot - SPIGOT-7159: Better modifier resolution
+                    int i = (Integer) state.getValue(StemBlock.AGE);
+
+                    if (i < 7) {
+                        state = (BlockState) state.setValue(StemBlock.AGE, i + 1);
+                        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, state, 2); // CraftBukkit
+                    } else {
+                        Direction enumdirection = Direction.Plane.HORIZONTAL.getRandomDirection(random);
+                        BlockPos blockposition1 = pos.relative(enumdirection);
+                        BlockState iblockdata1 = level.getBlockState(blockposition1.below());
+
+                        if (level.getBlockState(blockposition1).isAir() && (iblockdata1.is(Blocks.FARMLAND) || iblockdata1.is(BlockTags.DIRT))) {
+                            Registry<Block> iregistry = net.minecraft.core.registries.BuiltInRegistries.BLOCK;
+                            Optional<Block> optional = iregistry.getOptional(this.fruit);
+                            Optional<Block> optional1 = iregistry.getOptional(this.attachedStem);
+
+                            if (optional.isPresent() && optional1.isPresent()) {
+                                // CraftBukkit start
+                                if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, blockposition1, ((Block) optional.get()).defaultBlockState(), 2)) {
+                                    return;
+                                }
+                                // CraftBukkit end
+                                // Now that we know that the crop will grow... is the next stage the crop's max age? If yes, we are going to check if the farm land is moist!
+                                if (isCurrentFarmlandStateMoist && !FarmBlock.isNearWater(level, farmlandBelowTheCurrentBlock)) {
+                                    // Whoops, farm land ain't moist!
+                                    // From FarmBlock, set the moisture to 0
+                                    org.bukkit.craftbukkit.event.CraftEventFactory.handleMoistureChangeEvent(level, farmlandBelowTheCurrentBlock, (BlockState) farmlandBelowTheCurrentBlockData.setValue(FarmBlock.MOISTURE, 0), 2); // CraftBukkit
+                                }
+                                level.setBlockAndUpdate(pos, (BlockState) ((Block) optional1.get()).defaultBlockState().setValue(HorizontalDirectionalBlock.FACING, enumdirection));
+                            }
+                        }
+                    }
+                }
+            } else {
             float growthSpeed = CropBlock.getGrowthSpeed(this, level, pos);
             if (random.nextFloat() < ((this == Blocks.PUMPKIN_STEM ? level.spigotConfig.pumpkinModifier : level.spigotConfig.melonModifier) / (100.0F * (Math.floor((25.0F / growthSpeed) + 1))))) { // Spigot - SPIGOT-7159: Better modifier resolution
                 int ageValue = state.getValue(AGE);
@@ -95,6 +_,8 @@
                 }
             }
         }
+        }
+        // SparklyPaper end
     }
 
     @Override
