--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -118,7 +_,7 @@
         .add(new ExplosionParticleInfo(ParticleTypes.POOF, 0.5F, 1.0F))
         .add(new ExplosionParticleInfo(ParticleTypes.SMOKE, 1.0F, 1.0F))
         .build();
-    public final List<TickingBlockEntity> blockEntityTickers = Lists.newArrayList();
+    public final net.sparklypower.sparklypaper.BlockEntityTickersList blockEntityTickers = new net.sparklypower.sparklypaper.BlockEntityTickersList(); // SparklyPaper - optimize block entity removals
     protected final CollectingNeighborUpdater neighborUpdater;
     private final List<TickingBlockEntity> pendingBlockEntityTickers = Lists.newArrayList();
     private boolean tickingBlockEntities;
@@ -146,6 +_,7 @@
 
     // CraftBukkit start
     public final io.papermc.paper.antixray.ChunkPacketBlockController chunkPacketBlockController; // Paper - Anti-Xray
+    public net.sparklypower.sparklypaper.configs.SparklyPaperConfig.SparklyPaperWorldConfig sparklyPaperConfig; // SparklyPaper - config files
     private final CraftWorld world;
     public org.bukkit.generator.@Nullable ChunkGenerator generator;
 
@@ -852,6 +_,7 @@
         final org.bukkit.NamespacedKey worldKey = CraftNamespacedKey.fromMinecraft(dimension.identifier()); // Paper
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(bukkitName, worldKey); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
+        this.sparklyPaperConfig = net.sparklypower.sparklypaper.configs.SparklyPaperConfigUtils.INSTANCE.getWorldSettings(((net.minecraft.world.level.storage.ServerLevelData) levelData).getLevelName()); // SparklyPaper - config files
         this.generator = generator;
         this.world = new CraftWorld((ServerLevel) this, worldKey, biomeProvider, environment);
 
@@ -1491,14 +_,10 @@
         boolean tickBlockEntities = this.tickRateManager().runsNormally();
 
         int tickedEntities = 0; // Paper - rewrite chunk system
-        // Paper start - Fix MC-117075 use removeAll
-        final it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<@Nullable TickingBlockEntity> toRemove = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>();
-        toRemove.add(null);
-        for (int tickerIndex = 0; tickerIndex < this.blockEntityTickers.size(); tickerIndex++) {
+        for (int tickerIndex = 0; tickerIndex < this.blockEntityTickers.size(); tickerIndex++) { // Paper - Fix MC-117075 use removeAll
             final TickingBlockEntity ticker = this.blockEntityTickers.get(tickerIndex);
-        // Paper end - Fix MC-117075 use removeAll
             if (ticker.isRemoved()) {
-                toRemove.add(ticker); // Paper - Fix MC-117075 use removeAll
+                this.blockEntityTickers.markAsRemoved(tickerIndex); // SparklyPaper - optimize block entity removals
             } else if (tickBlockEntities && this.shouldTickBlocksAt(ticker.getPos())) {
                 ticker.tick();
                 // Paper start - rewrite chunk system
@@ -1509,7 +_,7 @@
             }
         }
 
-        this.blockEntityTickers.removeAll(toRemove); // Paper - Fix MC-117075 use removeAll
+        this.blockEntityTickers.removeMarkedEntries(); // SparklyPaper - optimize block entity removals
         this.tickingBlockEntities = false;
     }
 
