From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Tue, 21 Mar 2023 01:17:52 +0000
Subject: [PATCH] More note block events

From Haricot, thanks Kezz! https://github.com/kezz/Haricot/blob/dffbe897cd6db773c06cfad6304e20b54c7aa980/patches/server/0005-More-note-block-events.patch

diff --git a/src/main/java/net/minecraft/world/level/block/NoteBlock.java b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
index 50ead76879398222a76f26c36e98800d2d1af167..a551bd71a2c1c2096296c35862f0e79c781a41ce 100644
--- a/src/main/java/net/minecraft/world/level/block/NoteBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
@@ -31,6 +31,12 @@ import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import net.minecraft.world.level.block.state.properties.NoteBlockInstrument;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.BlockHitResult;
+// Haricot start
+import ml.beancraft.haricot.event.block.NoteBlockPlaceEvent;
+import ml.beancraft.haricot.event.block.NoteBlockUpdateEvent;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.block.impl.CraftNote;
+// Haricot end
 
 public class NoteBlock extends Block {
 
@@ -62,14 +68,35 @@ public class NoteBlock extends Block {
 
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext ctx) {
-        return this.setInstrument(ctx.getLevel(), ctx.getClickedPos(), this.defaultBlockState());
+        // Haricot start - note block api
+        NoteBlockPlaceEvent event = new NoteBlockPlaceEvent(new CraftBlock(ctx.getLevel(), ctx.getClickedPos()),
+                new CraftNote(this.getStateDefinition().any()), ctx.getItemInHand().asBukkitMirror());
+        event.callEvent();
+
+        return this.getStateDefinition().any().setValue(INSTRUMENT, NoteBlockInstrument.values()[event.getData().getInstrument().getType()])
+            .setValue(NOTE, (int) event.getData().getNote().getId())
+            .setValue(POWERED, event.getData().isPowered());
+        // Haricot end
     }
 
     @Override
     public BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor world, BlockPos pos, BlockPos neighborPos) {
+        // Haricot start - note block api
         boolean flag = NoteBlock.isFeatureFlagEnabled(world) ? direction.getAxis() == Direction.Axis.Y : direction == Direction.DOWN;
 
-        return flag ? this.setInstrument(world, pos, state) : super.updateShape(state, direction, neighborState, world, pos, neighborPos);
+        BlockState newBlockState = flag ? this.setInstrument(world, pos, state) : super.updateShape(state, direction, neighborState, world, pos, neighborPos);
+
+        NoteBlockUpdateEvent event = new NoteBlockUpdateEvent(new CraftBlock(world, pos), new CraftNote(state), new CraftNote(newBlockState));
+        event.callEvent();
+
+        if (!event.isCancelled()) {
+            return state.setValue(INSTRUMENT, NoteBlockInstrument.values()[event.getNewData().getInstrument().getType()])
+                    .setValue(NOTE, (int) event.getNewData().getNote().getId())
+                    .setValue(POWERED, event.getNewData().isPowered());
+        } else {
+            return state;
+        }
+        // Haricot end
     }
 
     @Override
@@ -82,7 +109,22 @@ public class NoteBlock extends Block {
                 state = world.getBlockState(pos); // CraftBukkit - SPIGOT-5617: update in case changed in event
             }
 
-            world.setBlock(pos, (BlockState) state.setValue(NoteBlock.POWERED, flag1), 3);
+            // Haricot start - note block api
+            NoteBlockUpdateEvent event = new NoteBlockUpdateEvent(new CraftBlock(world, pos), new CraftNote(state),
+                    new CraftNote(state.setValue(NoteBlock.POWERED, flag1)));
+            event.callEvent();
+
+            if (!event.isCancelled()) {
+                world.setBlock(
+                        pos,
+                        state
+                                .setValue(NoteBlock.INSTRUMENT, NoteBlockInstrument.values()[event.getNewData().getInstrument().getType()])
+                                .setValue(NoteBlock.NOTE, (int) event.getNewData().getNote().getId())
+                                .setValue(NoteBlock.POWERED, flag1),
+                        3
+                );
+            }
+            // Haricot end
         }
 
     }
@@ -118,7 +160,22 @@ public class NoteBlock extends Block {
             state = (BlockState) state.cycle(NoteBlock.NOTE);
             world.setBlock(pos, state, 3);
             this.playNote(player, state, world, pos);
-            player.awardStat(Stats.TUNE_NOTEBLOCK);
+            // Haricot start - note block api
+            NoteBlockUpdateEvent event = new NoteBlockUpdateEvent(new CraftBlock(world, pos), new CraftNote(state),
+                    new CraftNote(state.cycle(NOTE)));
+            event.callEvent();
+
+            if (!event.isCancelled()) {
+                BlockState newData = state
+                        .setValue(INSTRUMENT, NoteBlockInstrument.values()[event.getNewData().getInstrument().getType()])
+                        .setValue(NOTE, (int) event.getNewData().getNote().getId())
+                        .setValue(POWERED, event.getNewData().isPowered());
+
+                world.setBlock(pos, newData, 3);
+                this.playNote(player, state, world, pos); // CraftBukkit
+                player.awardStat(Stats.TUNE_NOTEBLOCK);
+            }
+            // Haricot end
             return InteractionResult.CONSUME;
         }
     }
@@ -184,4 +241,4 @@ public class NoteBlock extends Block {
     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(NoteBlock.INSTRUMENT, NoteBlock.POWERED, NoteBlock.NOTE);
     }
-}
+}
\ No newline at end of file
