From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Sun, 12 Jan 2025 23:26:44 -0300
Subject: [PATCH] Add PreEntityShootBowEvent


diff --git a/src/main/java/net/minecraft/world/item/BowItem.java b/src/main/java/net/minecraft/world/item/BowItem.java
index bb593209c95c9cf1f9c5d52d52fab4a33ddbabcf..776c37284e9cff50fae6b937cb707e7ba7edc624 100644
--- a/src/main/java/net/minecraft/world/item/BowItem.java
+++ b/src/main/java/net/minecraft/world/item/BowItem.java
@@ -33,6 +33,11 @@ public class BowItem extends ProjectileWeaponItem {
             } else {
                 int i = this.getUseDuration(stack, user) - remainingUseTicks;
                 float f = getPowerForTime(i);
+                // SparklyPaper start - Add PreEntityShootBowEvent
+                net.sparklypower.sparklypaper.event.entity.PreEntityShootBowEvent event = new net.sparklypower.sparklypaper.event.entity.PreEntityShootBowEvent(player.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack), player.getUsedItemHand() == InteractionHand.MAIN_HAND ? org.bukkit.inventory.EquipmentSlot.HAND : org.bukkit.inventory.EquipmentSlot.OFF_HAND, f);
+                if (!event.callEvent())
+                    return false;
+                // SparklyPaper end
                 if ((double)f < 0.1) {
                     return false;
                 } else {
