From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Sat, 7 Aug 2021 13:36:22 -0300
Subject: [PATCH] Support complex components in written book builder


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
index a13e75797d612d79aa26c4891f66ecf0ce2bebaa..ebb643fc0477409d1efb4e9af59675045fa6b8bb 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
@@ -360,7 +360,7 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
         this.pages = pages.subList(0, Math.min(MAX_PAGES, pages.size())).stream().map(io.papermc.paper.adventure.PaperAdventure.LEGACY_SECTION_UXRC::serialize).collect(java.util.stream.Collectors.toList());
     }
 
-    static final class CraftMetaBookBuilder implements BookMetaBuilder {
+    static class CraftMetaBookBuilder implements BookMetaBuilder {
         private net.kyori.adventure.text.Component title = null;
         private net.kyori.adventure.text.Component author = null;
         private final List<net.kyori.adventure.text.Component> pages = new java.util.ArrayList<>();
@@ -397,6 +397,10 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
 
         @Override
         public BookMeta build() {
+            return this.build(title, author, pages);
+        }
+
+        protected BookMeta build(net.kyori.adventure.text.Component title, net.kyori.adventure.text.Component author, java.util.List<net.kyori.adventure.text.Component> pages) {
             return new CraftMetaBook(title, author, pages);
         }
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java
index 0cf60eb9b6ba1a79c9b603c4349debd478101f9a..e09b0d61af941dc6d8f7e42ed40a6eafdd137314 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java
@@ -88,4 +88,25 @@ class CraftMetaBookSigned extends CraftMetaBook implements BookMeta {
         super.serialize(builder);
         return builder;
     }
+    
+    // Paper start - adventure
+    private CraftMetaBookSigned(net.kyori.adventure.text.Component title, net.kyori.adventure.text.Component author, java.util.List<net.kyori.adventure.text.Component> pages) {
+        super((org.bukkit.craftbukkit.inventory.CraftMetaItem) org.bukkit.Bukkit.getItemFactory().getItemMeta(Material.WRITABLE_BOOK));
+        this.title = title == null ? null : io.papermc.paper.adventure.PaperAdventure.LEGACY_SECTION_UXRC.serialize(title);
+        this.author = author == null ? null : io.papermc.paper.adventure.PaperAdventure.LEGACY_SECTION_UXRC.serialize(author);
+        this.pages = io.papermc.paper.adventure.PaperAdventure.asJson(pages.subList(0, Math.min(MAX_PAGES, pages.size())));
+    }
+
+    static final class CraftMetaBookSignedBuilder extends CraftMetaBookBuilder {
+        @Override
+        protected BookMeta build(net.kyori.adventure.text.Component title, net.kyori.adventure.text.Component author, java.util.List<net.kyori.adventure.text.Component> pages) {
+            return new CraftMetaBookSigned(title, author, pages);
+        }
+    }
+
+    @Override
+    public BookMetaBuilder toBuilder() {
+        return new CraftMetaBookSignedBuilder();
+    }
+    // Paper end
 }
