From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Fri, 17 Jul 2026 18:38:10 -0300
Subject: [PATCH] Add per-entity tracking range overrides


diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 7f0df1c26f93d53b5925c564374f3e5fa6d966b3..2460f3f63f37b74deaaf9f8fe42f0dfbc49330e4 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -995,6 +995,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             EntityType<?> type = entity.getType();
             int range = type.clientTrackingRange() * 16;
             range = org.spigotmc.TrackingRange.getEntityTrackingRange(entity, range); // Spigot
+            if (entity.trackingRangeOverride != -1) range = entity.trackingRangeOverride; // SparklyPaper - per-entity tracking range override
             if (range != 0) {
                 int updateInterval = type.updateInterval();
                 if (this.entityMap.containsKey(entity.getId())) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 1530810cdb74810e37e3e22ce4fcda4ee7511055..9323beb98dc2a86e67111c089aa4c7f0125bdfe1 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -382,6 +382,7 @@ public abstract class Entity
     private final int despawnTime; // Paper - entity despawn time limit
     public int totalEntityAge; // Paper - age-like counter for all entities
     public final io.papermc.paper.entity.activation.ActivationType activationType = io.papermc.paper.entity.activation.ActivationType.activationTypeFor(this); // Paper - EAR 2/tracking ranges
+    public int trackingRangeOverride = -1; // SparklyPaper - per-entity tracking range override
     // Paper start - EAR 2
     public final boolean defaultActivationState;
     public long activatedTick = Integer.MIN_VALUE;
