From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Fri, 6 Feb 2026 16:04:58 -0300
Subject: [PATCH] Add WorldTickStartEvent and WorldTickEndEvent


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index e1c60bccc68c724ce61a790fc9e23f875617a52c..a12171f1d9ecbccfe154194f8710db469c95eda4 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1831,11 +1831,15 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                         ca.spottedleaf.moonrise.common.util.TickThread.ServerLevelTickThread currentThread = (ca.spottedleaf.moonrise.common.util.TickThread.ServerLevelTickThread) Thread.currentThread();
                         currentThread.currentlyTickingServerLevel = serverLevel;
 
+                        new net.sparklypower.sparklypaper.event.world.WorldTickStartEvent(serverLevel.getWorld(), this.tickCount).callEvent();
+
                         long i = Util.getNanos(); // SparklyPaper - track world's MSPT
                         serverLevel.tick(hasTimeLeft);
                         // SparklyPaper start - track world's MSPT
                         long j = Util.getNanos() - i;
 
+                        new net.sparklypower.sparklypaper.event.world.WorldTickEndEvent(serverLevel.getWorld(), this.tickCount, j).callEvent();
+                        
                         // These are from the "tickServer" function
                         serverLevel.tickTimes5s.add(this.tickCount, j);
                         serverLevel.tickTimes10s.add(this.tickCount, j);
