From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Fri, 17 Jul 2026 18:38:10 -0300
Subject: [PATCH] Add per-entity tracking range overrides


diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 7e9361407042ddaabe3f66ee8957ae84b6ef2317..391bd436badb099f60ebdeed7fbc0377666814ab 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -995,6 +995,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             EntityType<?> type = entity.getType();
             int range = type.clientTrackingRange() * 16;
             range = org.spigotmc.TrackingRange.getEntityTrackingRange(entity, range); // Spigot
+            if (entity.trackingRangeOverride != -1) range = entity.trackingRangeOverride; // SparklyPaper - per-entity tracking range override
             if (range != 0) {
                 int updateInterval = type.updateInterval();
                 if (this.entityMap.containsKey(entity.getId())) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index d958c4717739e588c5f8985dec5d215c745f49fc..6863cf758ca591a9a1ce607633f5385cfd7eccd2 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -383,6 +383,7 @@ public abstract class Entity
     private final int despawnTime; // Paper - entity despawn time limit
     public int totalEntityAge; // Paper - age-like counter for all entities
     public final io.papermc.paper.entity.activation.ActivationType activationType = io.papermc.paper.entity.activation.ActivationType.activationTypeFor(this); // Paper - EAR 2/tracking ranges
+    public int trackingRangeOverride = -1; // SparklyPaper - per-entity tracking range override
     // Paper start - EAR 2
     public final boolean defaultActivationState;
     public long activatedTick = Integer.MIN_VALUE;
