From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Fri, 6 Feb 2026 16:04:58 -0300
Subject: [PATCH] Add WorldTickStartEvent and WorldTickEndEvent


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 1461d1fac7c9ccc3a7e4fade3919843732f819cf..6c41f7697c2c3f1cd8035b8d9d1ecfbd289cf9e4 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1832,11 +1832,15 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                         ca.spottedleaf.moonrise.common.util.TickThread.ServerLevelTickThread currentThread = (ca.spottedleaf.moonrise.common.util.TickThread.ServerLevelTickThread) Thread.currentThread();
                         currentThread.currentlyTickingServerLevel = level;
 
+                        new net.sparklypower.sparklypaper.event.world.WorldTickStartEvent(level.getWorld(), this.tickCount).callEvent();
+
                         long i = Util.getNanos(); // SparklyPaper - track world's MSPT
                         level.tick(haveTime);
                         // SparklyPaper start - track world's MSPT
                         long j = Util.getNanos() - i;
 
+                        new net.sparklypower.sparklypaper.event.world.WorldTickEndEvent(level.getWorld(), this.tickCount, j).callEvent();
+                        
                         // These are from the "tickServer" function
                         level.tickTimes5s.add(this.tickCount, j);
                         level.tickTimes10s.add(this.tickCount, j);
