From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Thu, 25 Sep 2025 21:17:14 -0300
Subject: [PATCH] Parallel World Ticking


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
index eb7c5ba7168ba4903e93938dd8671227194bec99..40cde921840ac43e9f0e22ef61d3989009964fa3 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -1153,7 +1153,7 @@ public final class ChunkHolderManager {
         if (changedFullStatus.isEmpty()) {
             return;
         }
-        if (!TickThread.isTickThread()) {
+        if (!TickThread.isTickThreadFor(world)) { // SparklyPaper - parallel world ticking
             // These will be handled on the next ServerChunkCache$MainThreadExecutor#pollTask, as it runs the distance manager update
             // which will invoke processTicketUpdates
             this.offThreadPendingFullLoadUpdate.addAll(changedFullStatus);
@@ -1175,7 +1175,7 @@ public final class ChunkHolderManager {
 
     // note: never call while inside the chunk system, this will absolutely break everything
     public void processUnloads() {
-        TickThread.ensureTickThread("Cannot unload chunks off-main");
+        TickThread.ensureTickThread(world, "Cannot unload chunks off-main"); // SparklyPaper - parallel world ticking
 
         if (BLOCK_TICKET_UPDATES.get() == Boolean.TRUE) {
             throw new IllegalStateException("Cannot unload chunks recursively");
@@ -1444,7 +1444,7 @@ public final class ChunkHolderManager {
         if (BLOCK_TICKET_UPDATES.get() == Boolean.TRUE) {
             throw new IllegalStateException("Cannot update ticket level while unloading chunks or updating entity manager");
         }
-        final boolean isTickThread = TickThread.isTickThread();
+        final boolean isTickThread = TickThread.isTickThreadFor(world);
 
         if (!PlatformHooks.get().allowAsyncTicketUpdates() && isTickThread) {
             TickThread.ensureTickThread("Cannot asynchronously process ticket updates");
diff --git a/io/papermc/paper/redstone/RedstoneWireTurbo.java b/io/papermc/paper/redstone/RedstoneWireTurbo.java
index f7820568b78f630e4d234d3dfb36c2167149777a..0ed1b717b318caceee913bbdd96e9bb8f813ad9e 100644
--- a/io/papermc/paper/redstone/RedstoneWireTurbo.java
+++ b/io/papermc/paper/redstone/RedstoneWireTurbo.java
@@ -827,14 +827,15 @@ public final class RedstoneWireTurbo {
         j = getMaxCurrentStrength(upd, j);
         int l = 0;
 
-        wire.shouldSignal = false;
+        net.sparklypower.sparklypaper.threadedregions.GenericWorldData worldData = net.sparklypower.sparklypaper.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
+        worldData.shouldSignal = false;
         // Unfortunately, World.isBlockIndirectlyGettingPowered is complicated,
         // and I'm not ready to try to replicate even more functionality from
         // elsewhere in Minecraft into this accelerator.  So sadly, we must
         // suffer the performance hit of this very expensive call.  If there
         // is consistency to what this call returns, we may be able to cache it.
         final int k = worldIn.getBestNeighborSignal(upd.self);
-        wire.shouldSignal = true;
+        worldData.shouldSignal = true;
 
         // The variable 'k' holds the maximum redstone power value of any adjacent blocks.
         // If 'k' has the highest level of all neighbors, then the power level of this
diff --git a/net/minecraft/core/dispenser/DispenseItemBehavior.java b/net/minecraft/core/dispenser/DispenseItemBehavior.java
index c650f26d3ecdeb94a4d0f40637c07c59a6cbdfa6..9eec823c117480f38ca73dd334f65554fb8eee4b 100644
--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -308,8 +308,8 @@ public interface DispenseItemBehavior {
                     capturedBlockStates = new java.util.ArrayList<>(level.capturedBlockStates.values());
                     level.capturedBlockStates.clear();
 
-                    treeType = net.minecraft.world.level.block.SaplingBlock.treeType;
-                    net.minecraft.world.level.block.SaplingBlock.treeType = null;
+                    treeType = net.minecraft.world.level.block.SaplingBlock.treeTypeRT.get(); // SparklyPaper - parallel world ticking
+                    net.minecraft.world.level.block.SaplingBlock.treeTypeRT.set(null); // SparklyPaper - parallel world ticking
                 }
                 // Paper end
                 // CraftBukkit start
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index ead67407482c8d9104fb17e9269de279e5990ac6..20ecb7919c329f760d0fae70e67f93f2615fe2c1 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -420,6 +420,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         };
     }
     // Paper end - improve tick loop
+    public java.util.concurrent.Semaphore serverLevelTickingSemaphore = null; // SparklyPaper - parallel world ticking
+    public final net.sparklypower.sparklypaper.threadedregions.GenericWorldData fallbackWorldData = new net.sparklypower.sparklypaper.threadedregions.GenericWorldData(); // SparklyPaper - parallel world ticking
+
     // Paper start - rewrite chunk system
     private volatile Throwable chunkSystemCrash;
 
@@ -1793,34 +1796,55 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.updateEffectiveRespawnData();
 
         this.isIteratingOverLevels = true; // Paper - Throw exception on world create while being ticked
+        // SparklyPaper start - parallel world ticking
+        java.util.ArrayDeque<java.util.concurrent.Future<ServerLevel>> tasks = new java.util.ArrayDeque<>();
+        try {
         for (ServerLevel level : this.getAllLevels()) {
             level.hasPhysicsEvent = org.bukkit.event.block.BlockPhysicsEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - BlockPhysicsEvent
             level.hasEntityMoveEvent = io.papermc.paper.event.entity.EntityMoveEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - Add EntityMoveEvent
             level.updateLagCompensationTick(); // Paper - lag compensation
-            net.minecraft.world.level.block.entity.HopperBlockEntity.skipHopperEvents = level.paperConfig().hopper.disableMoveEvent || org.bukkit.event.inventory.InventoryMoveItemEvent.getHandlerList().getRegisteredListeners().length == 0; // Paper - Perf: Optimize Hoppers
+            level.regionizedWorldData.skipHopperEvents = level.paperConfig().hopper.disableMoveEvent || org.bukkit.event.inventory.InventoryMoveItemEvent.getHandlerList().getRegisteredListeners().length == 0; // Paper - Perf: Optimize Hoppers
             profiler.push(() -> level + " " + level.dimension().identifier());
             profiler.push("tick");
 
-            try {
-                long i = Util.getNanos(); // SparklyPaper - track world's MSPT
-                level.tick(haveTime);
-                // SparklyPaper start - track world's MSPT
-                long j = Util.getNanos() - i;
-                // These are from the "tickServer" function
-                level.tickTimes5s.add(this.tickCount, j);
-                level.tickTimes10s.add(this.tickCount, j);
-                level.tickTimes1m.add(this.tickCount, j);
-                // SparklyPaper end
-            } catch (Throwable t) {
-                CrashReport report = CrashReport.forThrowable(t, "Exception ticking world");
-                level.fillReportDetails(report);
-                throw new ReportedException(report);
-            }
+            serverLevelTickingSemaphore.acquire();
+            tasks.add(
+                level.tickExecutor.submit(() -> {
+                    try {
+                        ca.spottedleaf.moonrise.common.util.TickThread.ServerLevelTickThread currentThread = (ca.spottedleaf.moonrise.common.util.TickThread.ServerLevelTickThread) Thread.currentThread();
+
+                        long i = Util.getNanos(); // SparklyPaper - track world's MSPT
+                        level.tick(haveTime);
+                        // SparklyPaper start - track world's MSPT
+                        long j = Util.getNanos() - i;
+
+                        // These are from the "tickServer" function
+                        level.tickTimes5s.add(this.tickCount, j);
+                        level.tickTimes10s.add(this.tickCount, j);
+                        level.tickTimes1m.add(this.tickCount, j);
+                        // SparklyPaper end
+                    } catch (Throwable t) {
+                        CrashReport report = CrashReport.forThrowable(t, "Exception ticking world");
+                        level.fillReportDetails(report);
+                        throw new ReportedException(report);
+                    } finally {
+                        serverLevelTickingSemaphore.release();
+                        level.explosionDensityCache.clear(); // Paper - Optimize explosions
+                    }
+                }, level)
+            );
 
             profiler.pop();
             profiler.pop();
-            level.explosionDensityCache.clear(); // Paper - Optimize explosions
         }
+
+        while (!tasks.isEmpty()) {
+            tasks.pop().get();
+        }
+        } catch (java.lang.InterruptedException | java.util.concurrent.ExecutionException e) {
+        throw new RuntimeException(e); // Propagate exception
+        }
+        // SparklyPaper end
         this.isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
 
         profiler.popPush("connection");
@@ -1911,6 +1935,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         Map<ResourceKey<Level>, ServerLevel> oldLevels = this.levels;
         Map<ResourceKey<Level>, ServerLevel> newLevels = Maps.newLinkedHashMap(oldLevels);
         newLevels.remove(level.dimension());
+        level.tickExecutor.shutdown(); // SparklyPaper - parallel world ticking (We remove it in here instead of ServerLevel.close() because ServerLevel.close() is never called!)
         this.levels = Collections.unmodifiableMap(newLevels);
     }
     // CraftBukkit end
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index c6255f79f369c8618282307db5ae0c9c7a49fce9..1e24d164d4e3577eb6161408c52a6005336bd791 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -243,6 +243,10 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
         net.sparklypower.sparklypaper.SparklyPaperCommands.INSTANCE.registerCommands(this);
         // SparklyPaper end
+        // SparklyPaper start - parallel world ticking start
+        serverLevelTickingSemaphore = new java.util.concurrent.Semaphore(net.sparklypower.sparklypaper.configs.SparklyPaperConfigUtils.INSTANCE.getConfig().getParallelWorldTicking().getThreads());
+        DedicatedServer.LOGGER.info("Using " + serverLevelTickingSemaphore.availablePermits() + " permits for parallel world ticking"); // SparklyPaper - parallel world ticking
+        // SparklyPaper end
         // Paper start - initialize global and world-defaults configuration
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index c4cfb53e145d69666ffc27ef4ee0c50b54eb9143..48dbda1457c14f411de8d76bbb71be2db9b6bb29 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -179,7 +179,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
             // call mid-tick tasks for chunk system
             if ((i & 7) == 0) {
-                ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.level.getServer()).moonrise$executeMidTickTasks();
+                // ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.level.getServer()).moonrise$executeMidTickTasks(); // SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks)
                 continue;
             }
         }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index b09d608074628fafaeb6c543fca0f2e3cba5650d..8c6ab1590ec38c80960411be6b991979b0ee8733 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -197,7 +197,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     private final ServerChunkCache chunkSource;
     private final MinecraftServer server;
     public final io.papermc.paper.world.saveddata.PaperLevelOverrides serverLevelData; // Paper - type
-    private final EntityTickList entityTickList = new EntityTickList();
+    final EntityTickList entityTickList = new EntityTickList(this); // SparklyPaper - parallel world ticking
     private final ServerWaypointManager waypointManager;
     private EnvironmentAttributeSystem environmentAttributes;
     // Paper - rewrite chunk system
@@ -222,7 +222,9 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     private final StructureManager structureManager;
     private final StructureCheck structureCheck;
     private final boolean tickTime;
-    private final LevelDebugSynchronizers debugSynchronizers = new LevelDebugSynchronizers(this);
+    final LevelDebugSynchronizers debugSynchronizers = new LevelDebugSynchronizers(this);
+    public final java.util.concurrent.ExecutorService tickExecutor; // SparklyPaper - parallel world ticking
+    public final net.sparklypower.sparklypaper.threadedregions.GenericWorldData regionizedWorldData = new net.sparklypower.sparklypaper.threadedregions.GenericWorldData(); // SparklyPaper - parallel world ticking
 
     // CraftBukkit start
     private final ResourceKey<LevelStem> typeKey;
@@ -750,6 +752,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         this.chunkDataController = new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.ChunkDataController((ServerLevel)(Object)this, this.chunkTaskScheduler);
         // Paper end - rewrite chunk system
         this.getCraftServer().addWorld(this.getWorld()); // CraftBukkit
+        this.tickExecutor = java.util.concurrent.Executors.newSingleThreadExecutor(new net.sparklypower.sparklypaper.ServerLevelTickExecutorThreadFactory(getWorld().getName())); // SparklyPaper - parallel world ticking
     }
 
     // Paper start
@@ -1369,7 +1372,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         }
         // Paper start - rewrite chunk system
         if ((++this.tickedBlocksOrFluids & 7L) != 0L) {
-            ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
+            // ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks(); // SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks)
         }
         // Paper end - rewrite chunk system
 
@@ -1382,7 +1385,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         }
         // Paper start - rewrite chunk system
         if ((++this.tickedBlocksOrFluids & 7L) != 0L) {
-            ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
+            // ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks(); // SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks)
         }
         // Paper end - rewrite chunk system
 
@@ -1629,6 +1632,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
 
     private void addPlayer(final ServerPlayer player) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot add player off-main"); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         Entity existing = this.getEntity(player.getUUID());
         if (existing != null) {
             LOGGER.warn("Force-added player with duplicate UUID {}", player.getUUID());
@@ -1641,7 +1645,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     // CraftBukkit start
     private boolean addEntity(final Entity entity, final org.bukkit.event.entity.CreatureSpawnEvent.@Nullable SpawnReason spawnReason) {
-        org.spigotmc.AsyncCatcher.catchOp("entity add"); // Spigot
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot add entity off-main"); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         entity.generation = false; // Paper - Don't fire sync event during generation; Reset flag if it was added during a ServerLevel generation process
         // Paper start - extra debug info
         if (entity.valid) {
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 620833d07c1e6dda7e1e6bc4ed6aff9f82628b57..04bbae4a61bafa8f2369ec397183fcf208780c3a 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -483,6 +483,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     public @Nullable Long maxConcurrentChunkLoadsOverride = null;
     public @Nullable Long maxConcurrentChunkGeneratesOverride = null;
     // SparklyPaper end
+    public boolean hasTickedAtLeastOnceInNewWorld = false; // SparklyPaper - parallel world ticking (fixes bug in DreamResourceReset where the inventory is opened AFTER the player has changed worlds, if you click with the quick tp torch in a chest, because the inventory is opened AFTER the player has teleported)
 
     public ServerPlayer(final MinecraftServer server, final ServerLevel level, final GameProfile gameProfile, final ClientInformation clientInformation) {
         super(level, gameProfile);
@@ -731,6 +732,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
     @Override
     public void tick() {
+        hasTickedAtLeastOnceInNewWorld = true; // SparklyPaper - parallel world ticking (see: PARALLEL_NOTES.md - Opening an inventory after a world switch)
         // CraftBukkit start
         if (this.joining) {
             this.joining = false;
@@ -1551,6 +1553,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
         ServerLevel newLevel = transition.newLevel();
         ServerLevel oldLevel = this.level();
+
         ResourceKey<Level> lastDimension = oldLevel.dimension();
         // CraftBukkit start
         ResourceKey<net.minecraft.world.level.dimension.LevelStem> oldLevelTypeKey = oldLevel.getTypeKey();
@@ -1587,6 +1590,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             return this;
         }
 
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureOnlyTickThread("Cannot change dimension of a player off-main, from world " + oldLevel.getWorld().getName() + " to world " + newLevel.getWorld().getName()); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         this.isChangingDimension = true;
         LevelData levelData = newLevel.getLevelData();
         this.connection.send(new ClientboundRespawnPacket(this.createCommonSpawnInfo(newLevel), ClientboundRespawnPacket.KEEP_ALL_DATA));
@@ -1917,6 +1921,12 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
             return OptionalInt.empty();
         } else {
+            // SparklyPaper start - parallel world ticking (see: PARALLEL_NOTES.md - Opening an inventory after a world switch)
+            if (!hasTickedAtLeastOnceInNewWorld) {
+                MinecraftServer.LOGGER.warn("Ignoring request to open container " + menu + " because we haven't ticked in the current world yet!", new Throwable());
+                return OptionalInt.empty();
+            }
+            // SparklyPaper end
             // CraftBukkit start
             this.containerMenu = menu; // Moved up
             if (!this.isImmobile()) {
@@ -2012,6 +2022,11 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
     @Override
     public void closeContainer(org.bukkit.event.inventory.InventoryCloseEvent.Reason reason) {
+        // SparklyPaper start - parallel world ticking (debugging)
+        if (net.sparklypower.sparklypaper.configs.SparklyPaperConfigUtils.INSTANCE.getLogContainerCreationStacktraces()) {
+            MinecraftServer.LOGGER.warn("Closing " + this.getBukkitEntity().getName() + " inventory that was created at", this.containerMenu.containerCreationStacktrace);
+        }
+        // SparklyPaper end
         org.bukkit.craftbukkit.event.CraftEventFactory.handleInventoryCloseEvent(this, reason); // CraftBukkit
         // Paper end - Inventory close reason
         this.connection.send(new ClientboundContainerClosePacket(this.containerMenu.containerId));
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index a7f589970a04e520eb5cbe6dfeb333bc9cd479a8..d81d3c39215af83d38e695da49a8fec3fe30dc71 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -116,7 +116,7 @@ public abstract class PlayerList {
     private static final SimpleDateFormat BAN_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z", Locale.ROOT);
     private final MinecraftServer server;
     private final List<ServerPlayer> players = new java.util.concurrent.CopyOnWriteArrayList(); // CraftBukkit - ArrayList -> CopyOnWriteArrayList: Iterator safety
-    private final Map<UUID, ServerPlayer> playersByUUID = Maps.newHashMap();
+    private final Map<UUID, ServerPlayer> playersByUUID = Maps.newHashMap(); // SparklyPaper - parallel world ticking (we don't need to replace the original map because we never iterate on top of this map)
     private final Map<NameAndId, PermissionLevel> playerPermissions = Maps.newHashMap();
     private final UserBanList bans;
     private final IpBanList ipBans;
@@ -134,7 +134,7 @@ public abstract class PlayerList {
 
     // CraftBukkit start
     private org.bukkit.craftbukkit.CraftServer cserver;
-    private final Map<String,ServerPlayer> playersByName = new java.util.HashMap<>();
+    private final Map<String,ServerPlayer> playersByName = new java.util.HashMap<>(); // SparklyPaper - parallel world ticking (we don't need to replace the original map because we never iterate on top of this map)
     public @Nullable String collideRuleTeamName; // Paper - Configurable player collision
 
     public PlayerList(
@@ -158,6 +158,7 @@ public abstract class PlayerList {
     abstract public void loadAndSaveFiles(); // Paper - fix converting txt to json file; moved from DedicatedPlayerList constructor
 
     public void placeNewPlayer(final Connection connection, final ServerPlayer player, final CommonListenerCookie cookie) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureOnlyTickThread("Cannot place new player off-main"); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         player.isRealPlayer = true; // Paper
         player.loginTime = System.currentTimeMillis(); // Paper - Replace OfflinePlayer#getLastPlayed
         NameAndId gameProfile = player.nameAndId();
@@ -597,6 +598,7 @@ public abstract class PlayerList {
 
     // Paper start - respawn event
     public ServerPlayer respawn(final ServerPlayer serverPlayer, final boolean keepAllPlayerData, final Entity.RemovalReason removalReason, final org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureOnlyTickThread("Cannot respawn player off-main, respawning in world " + serverPlayer.level().getWorld().getName()); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         ServerPlayer.RespawnResult result = serverPlayer.findRespawnPositionAndUseSpawnBlock0(!keepAllPlayerData, TeleportTransition.DO_NOTHING, respawnReason);
         if (result == null) { // disconnected player during the respawn event
             return serverPlayer;
@@ -609,6 +611,7 @@ public abstract class PlayerList {
         serverPlayer.level().removePlayerImmediately(serverPlayer, removalReason);
         ServerLevel level = respawnInfo.newLevel();
         ServerPlayer player = serverPlayer; // Paper - TODO - recreate instance
+        player.hasTickedAtLeastOnceInNewWorld = false; // SparklyPaper - parallel world ticking (see: PARALLEL_NOTES.md - Opening an inventory after a world switch)
         player.connection = serverPlayer.connection;
         player.restoreFrom(serverPlayer, keepAllPlayerData);
         player.setId(serverPlayer.getId());
diff --git a/net/minecraft/server/waypoints/ServerWaypointManager.java b/net/minecraft/server/waypoints/ServerWaypointManager.java
index b9748a392c9f76e61a1b673085b4dd74c007c70d..330383e9005cd2244d576d5dfe1aef4fcafed437 100644
--- a/net/minecraft/server/waypoints/ServerWaypointManager.java
+++ b/net/minecraft/server/waypoints/ServerWaypointManager.java
@@ -21,14 +21,19 @@ public class ServerWaypointManager implements WaypointManager<WaypointTransmitte
     private final Table<ServerPlayer, WaypointTransmitter, WaypointTransmitter.Connection> connections = HashBasedTable.create();
     // Paper start - optimize ServerWaypointManager with locator bar disabled
     public boolean locatorBarEnabled; // Avoid a gamerule lookup for each player
+    // Paper end - optimize ServerWaypointManager with locator bar disabled
 
-    public ServerWaypointManager(net.minecraft.server.level.ServerLevel serverLevel) {
-        this.locatorBarEnabled = serverLevel.getGameRules().get(GameRules.LOCATOR_BAR);
+    // SparklyPaper start - parallel world ticking
+    private final net.minecraft.server.level.ServerLevel level;
+    public ServerWaypointManager(net.minecraft.server.level.ServerLevel level) {
+        this.level = level;
+        this.locatorBarEnabled = level.getGameRules().get(GameRules.LOCATOR_BAR); // Paper - optimize ServerWaypointManager with locator bar disabled
     }
-    // Paper end - optimize ServerWaypointManager with locator bar disabled
+    // SparklyPaper end
 
     @Override
     public void trackWaypoint(final WaypointTransmitter waypoint) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, "Cannot track waypoints off-main"); // SparklyPaper - parallel world ticking
         this.waypoints.add(waypoint);
 
         if (!this.locatorBarEnabled) return; // Paper - optimize ServerWaypointManager with locator bar disabled
@@ -39,6 +44,7 @@ public class ServerWaypointManager implements WaypointManager<WaypointTransmitte
 
     @Override
     public void updateWaypoint(final WaypointTransmitter waypoint) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, "Cannot update waypoints off-main"); // SparklyPaper - parallel world ticking
         if (!this.locatorBarEnabled) return; // Paper - optimize ServerWaypointManager with locator bar disabled
         if (this.waypoints.contains(waypoint)) {
             Map<ServerPlayer, WaypointTransmitter.Connection> playerConnection = Tables.transpose(this.connections).row(waypoint);
@@ -56,12 +62,14 @@ public class ServerWaypointManager implements WaypointManager<WaypointTransmitte
 
     @Override
     public void untrackWaypoint(final WaypointTransmitter waypoint) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, "Cannot untrack waypoints off-main"); // SparklyPaper - parallel world ticking
         this.connections.column(waypoint).forEach((player, connection) -> connection.disconnect());
         Tables.transpose(this.connections).row(waypoint).clear();
         this.waypoints.remove(waypoint);
     }
 
     public void addPlayer(final ServerPlayer player) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, "Cannot add player to waypoints off-main"); // SparklyPaper - parallel world ticking
         this.players.add(player);
 
         if (this.locatorBarEnabled) { // Paper - optimize ServerWaypointManager with locator bar disabled
@@ -76,6 +84,7 @@ public class ServerWaypointManager implements WaypointManager<WaypointTransmitte
     }
 
     public void updatePlayer(final ServerPlayer player) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, "Cannot update player for waypoints off-main"); // SparklyPaper - parallel world ticking
         if (!this.locatorBarEnabled) return; // Paper - optimize ServerWaypointManager with locator bar disabled
         Map<WaypointTransmitter, WaypointTransmitter.Connection> waypointConnections = this.connections.row(player);
         SetView<WaypointTransmitter> potentialWaypoints = Sets.difference(this.waypoints, waypointConnections.keySet());
@@ -90,6 +99,7 @@ public class ServerWaypointManager implements WaypointManager<WaypointTransmitte
     }
 
     public void removePlayer(final ServerPlayer player) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, "Cannot remove player from waypoints off-main"); // SparklyPaper - parallel world ticking
         this.connections.row(player).values().removeIf(connection -> {
             connection.disconnect();
             return true;
@@ -99,6 +109,7 @@ public class ServerWaypointManager implements WaypointManager<WaypointTransmitte
     }
 
     public void breakAllConnections() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, "Cannot break all waypoint connections off-main"); // SparklyPaper - parallel world ticking
         this.connections.values().forEach(WaypointTransmitter.Connection::disconnect);
         this.connections.clear();
     }
@@ -122,6 +133,7 @@ public class ServerWaypointManager implements WaypointManager<WaypointTransmitte
     }
 
     private void createConnection(final ServerPlayer player, final WaypointTransmitter waypoint) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, "Cannot create waypoint connections off-main"); // SparklyPaper - parallel world ticking
         if (player != waypoint) {
             if (isLocatorBarEnabledFor(player)) {
                 waypoint.makeWaypointConnectionWith(player).ifPresentOrElse(connection -> {
@@ -138,6 +150,7 @@ public class ServerWaypointManager implements WaypointManager<WaypointTransmitte
     }
 
     private void updateConnection(final ServerPlayer player, final WaypointTransmitter waypoint, final WaypointTransmitter.Connection connection) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, "Cannot update waypoint connection off-main"); // SparklyPaper - parallel world ticking
         if (player != waypoint) {
             if (isLocatorBarEnabledFor(player)) {
                 if (!connection.isBroken()) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index f23731f37c2120797688d5f595580039ca2d7780..213c3501d633b6fcf485e7c5d9c3e8353751f9e0 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3670,6 +3670,7 @@ public abstract class Entity
     }
 
     protected void handlePortal() {
+        if (true) return; // SparklyPaper - parallel world ticking (see issue #9, this is executed in the server level tick for non-player entities)
         if (this.level() instanceof ServerLevel level) {
             this.processPortalCooldown();
             if (this.portalProcess != null) {
@@ -4290,6 +4291,7 @@ public abstract class Entity
     }
 
     private @Nullable Entity teleportCrossDimension(final ServerLevel oldLevel, final ServerLevel newLevel, final TeleportTransition transition) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(oldLevel, "Cannot teleport entity to another world off-main, from world " + oldLevel.getWorld().getName() + " to world " + newLevel.getWorld().getName()); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         List<Entity> oldPassengers = this.getPassengers();
         List<Entity> newPassengers = new ArrayList<>(oldPassengers.size());
         this.ejectPassengers();
diff --git a/net/minecraft/world/inventory/AbstractContainerMenu.java b/net/minecraft/world/inventory/AbstractContainerMenu.java
index c223581e30414842ce24079b47c82639d36b7e73..b326d8f91e02dfbb49911f30b382a273373343d3 100644
--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -96,8 +96,14 @@ public abstract class AbstractContainerMenu {
 
     public void startOpen() {}
     // CraftBukkit end
+    public Throwable containerCreationStacktrace; // SparklyPaper - parallel world ticking (debugging)
 
     protected AbstractContainerMenu(final @Nullable MenuType<?> menuType, final int containerId) {
+        // SparklyPaper - parallel world ticking (debugging)
+        if (net.sparklypower.sparklypaper.configs.SparklyPaperConfigUtils.INSTANCE.getLogContainerCreationStacktraces()) {
+            this.containerCreationStacktrace = new Throwable();
+        }
+        // SparklyPaper end
         this.menuType = menuType;
         this.containerId = containerId;
     }
diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index e52d970069d2c4ebb7ca5138cc43ab50860c305e..7f1a2a01cce3db04aa6d7b69da580cf7a3d7f8de 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -404,8 +404,8 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
                 capturedBlockEntities = new java.util.ArrayList<>(level.capturedBlockEntities.values());
                 level.capturedBlockEntities.clear();
 
-                treeType = net.minecraft.world.level.block.SaplingBlock.treeType;
-                net.minecraft.world.level.block.SaplingBlock.treeType = null;
+                treeType = net.minecraft.world.level.block.SaplingBlock.treeTypeRT.get(); // SparklyPaper - parallel world ticking
+                net.minecraft.world.level.block.SaplingBlock.treeTypeRT.set(null); // SparklyPaper - parallel world ticking
             }
 
             DataComponentPatch newPatch = this.components.asPatch();
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index b95c82c4bd27dc16be4172acd7a778989cb69218..3641c9bd32c4d1a8f1e151ec1b343b3612b7a3d4 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -171,6 +171,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     }
     // Paper end - add paper world config
 
+    public io.papermc.paper.redstone.RedstoneWireTurbo turbo = new io.papermc.paper.redstone.RedstoneWireTurbo((net.minecraft.world.level.block.RedstoneWireBlock) net.minecraft.world.level.block.Blocks.REDSTONE_WIRE); // SparklyPaper - parallel world ticking (moved to world)
     public static @Nullable BlockPos lastPhysicsProblem; // Spigot
     public final Map<ServerExplosion.CacheKey, Float> explosionDensityCache = new java.util.HashMap<>(); // Paper - Optimize explosions
     public java.util.ArrayDeque<net.minecraft.world.level.block.RedstoneTorchBlock.Toggle> redstoneUpdateInfos; // Paper - Faster redstone torch rapid clock removal; Move from Map in BlockRedstoneTorch to here
@@ -1049,6 +1050,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
     @Override
     public boolean setBlock(final BlockPos pos, final BlockState blockState, final @Block.UpdateFlags int updateFlags, final int updateLimit) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread((ServerLevel)this, pos, "Updating block asynchronously"); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         // CraftBukkit start - tree generation
         if (this.captureTreeGeneration) {
             // Paper start - Protect Bedrock and End Portal/Frames from being destroyed
@@ -1506,7 +1508,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
                 ticker.tick();
                 // Paper start - rewrite chunk system
                 if ((++tickedEntities & 7) == 0) {
-                    ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel)(Level)(Object)this).moonrise$midTickTasks();
+                    // ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel)(Level)(Object)this).moonrise$midTickTasks(); // SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks)
                 }
                 // Paper end - rewrite chunk system
             }
@@ -1527,7 +1529,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
             entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
             // Paper end - Prevent block entity and entity crashes
         }
-        this.moonrise$midTickTasks(); // Paper - rewrite chunk system
+        // this.moonrise$midTickTasks(); // Paper - rewrite chunk system // SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks)
     }
 
     // Paper start - Option to prevent armor stands from doing entity lookups
@@ -1678,6 +1680,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
     @Override
     public @Nullable BlockEntity getBlockEntity(final BlockPos pos) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThreadOrAsyncThread((ServerLevel) this, "Cannot read world asynchronously"); // SparklyPaper - parallel world ticking
         // Paper start - Perf: Optimize capturedTileEntities lookup
         net.minecraft.world.level.block.entity.BlockEntity blockEntity;
         if (!this.capturedBlockEntities.isEmpty() && (blockEntity = this.capturedBlockEntities.get(pos)) != null) {
@@ -1694,6 +1697,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     }
 
     public void setBlockEntity(final BlockEntity blockEntity) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread((ServerLevel) this, "Cannot modify world asynchronously"); // SparklyPaper - parallel world ticking
         BlockPos pos = blockEntity.getBlockPos();
         if (this.isInValidBounds(pos)) {
             // CraftBukkit start
@@ -1766,6 +1770,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
     @Override
     public List<Entity> getEntities(final @Nullable Entity except, final AABB bb, final Predicate<? super Entity> selector) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread((ServerLevel)this, bb, "Cannot getEntities asynchronously"); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         Profiler.get().incrementCounter("getEntities");
         // Paper start - rewrite chunk system
         final List<Entity> ret = new java.util.ArrayList<>();
@@ -2124,8 +2129,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public abstract RecipeAccess recipeAccess();
 
     public BlockPos getBlockRandomPos(final int xo, final int yo, final int zo, final int yMask) {
-        this.randValue = this.randValue * 3 + 1013904223;
-        int val = this.randValue >> 2;
+        int val = this.random.nextInt() >> 2; // SparklyPaper - parallel world ticking
         return new BlockPos(xo + (val & 15), yo + (val >> 16 & yMask), zo + (val >> 8 & 15));
     }
 
diff --git a/net/minecraft/world/level/block/MushroomBlock.java b/net/minecraft/world/level/block/MushroomBlock.java
index c0779cb3c237e4399a53a7e6fc95bb2d88483bca..fd5a354e2739d250c8dfd9f6c83d230b06b9857f 100644
--- a/net/minecraft/world/level/block/MushroomBlock.java
+++ b/net/minecraft/world/level/block/MushroomBlock.java
@@ -83,7 +83,7 @@ public class MushroomBlock extends VegetationBlock implements BonemealableBlock
         }
 
         level.removeBlock(pos, false);
-        SaplingBlock.treeType = (this == Blocks.BROWN_MUSHROOM) ? org.bukkit.TreeType.BROWN_MUSHROOM : org.bukkit.TreeType.RED_MUSHROOM; // CraftBukkit
+        SaplingBlock.treeTypeRT.set((this == Blocks.BROWN_MUSHROOM) ? org.bukkit.TreeType.BROWN_MUSHROOM : org.bukkit.TreeType.RED_MUSHROOM); // CraftBukkit // SparklyPaper - parallel world ticking
         if (feature.get().value().place(level, level.getChunkSource().getGenerator(), random, pos)) {
             return true;
         }
diff --git a/net/minecraft/world/level/block/NetherFungusBlock.java b/net/minecraft/world/level/block/NetherFungusBlock.java
index fefb4e7a3bf922be16cf7fb6b70c0689867dd963..f94c1e514a57d8f273abd5c3b088b77118360e2a 100644
--- a/net/minecraft/world/level/block/NetherFungusBlock.java
+++ b/net/minecraft/world/level/block/NetherFungusBlock.java
@@ -64,9 +64,9 @@ public class NetherFungusBlock extends VegetationBlock implements BonemealableBl
             // CraftBukkit start
             .map((value) -> {
                 if (this == Blocks.WARPED_FUNGUS) {
-                    SaplingBlock.treeType = org.bukkit.TreeType.WARPED_FUNGUS;
+                    SaplingBlock.treeTypeRT.set(org.bukkit.TreeType.WARPED_FUNGUS); // SparklyPaper - parallel world ticking
                 } else if (this == Blocks.CRIMSON_FUNGUS) {
-                    SaplingBlock.treeType = org.bukkit.TreeType.CRIMSON_FUNGUS;
+                    SaplingBlock.treeTypeRT.set(org.bukkit.TreeType.CRIMSON_FUNGUS); // SparklyPaper - parallel world ticking
                 }
                 return value;
             })
diff --git a/net/minecraft/world/level/block/RedstoneWireBlock.java b/net/minecraft/world/level/block/RedstoneWireBlock.java
index ede64a0afbcdc41e60c3f2f5917c89f09b888433..0e155b92d95cb54ef50da9420c138648692bc1ee 100644
--- a/net/minecraft/world/level/block/RedstoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedstoneWireBlock.java
@@ -264,7 +264,7 @@ public class RedstoneWireBlock extends Block {
 
     // Paper start - Optimize redstone (Eigencraft)
     // The bulk of the new functionality is found in RedstoneWireTurbo.java
-    io.papermc.paper.redstone.RedstoneWireTurbo turbo = new io.papermc.paper.redstone.RedstoneWireTurbo(this);
+    // io.papermc.paper.redstone.RedstoneWireTurbo turbo = new io.papermc.paper.redstone.RedstoneWireTurbo(this); // SparklyPaper - parallel world ticking (moved to world)
 
     /*
      * Modified version of pre-existing updateSurroundingRedstone, which is called from
@@ -280,7 +280,7 @@ public class RedstoneWireBlock extends Block {
             if (orientation != null) {
                 source = pos.relative(orientation.getFront().getOpposite());
             }
-            turbo.updateSurroundingRedstone(worldIn, pos, state, source);
+            worldIn.turbo.updateSurroundingRedstone(worldIn, pos, state, source); // SparklyPaper - parallel world ticking
             return;
         }
         updatePowerStrength(worldIn, pos, state, orientation, blockAdded);
@@ -304,7 +304,7 @@ public class RedstoneWireBlock extends Block {
                 // [Space Walker] suppress shape updates and emit those manually to
                 // bypass the new neighbor update stack.
                 if (level.setBlock(pos, state, Block.UPDATE_KNOWN_SHAPE | Block.UPDATE_CLIENTS)) {
-                    turbo.updateNeighborShapes(level, pos, state);
+                    level.turbo.updateNeighborShapes(level, pos, state); // SparklyPaper - parallel world ticking
                 }
             }
         }
@@ -327,9 +327,10 @@ public class RedstoneWireBlock extends Block {
     }
 
     public int getBlockSignal(final Level level, final BlockPos pos) {
-        this.shouldSignal = false;
+        net.sparklypower.sparklypaper.threadedregions.GenericWorldData worldData = net.sparklypower.sparklypaper.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
+        worldData.shouldSignal = false;
         int blockSignal = level.getBestNeighborSignal(pos);
-        this.shouldSignal = true;
+        worldData.shouldSignal = true;
         return blockSignal;
     }
 
@@ -423,12 +424,14 @@ public class RedstoneWireBlock extends Block {
 
     @Override
     protected int getDirectSignal(final BlockState state, final BlockGetter level, final BlockPos pos, final Direction direction) {
-        return !this.shouldSignal ? Redstone.SIGNAL_MIN : state.getSignal(level, pos, direction);
+        net.sparklypower.sparklypaper.threadedregions.GenericWorldData worldData = net.sparklypower.sparklypaper.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
+        return !worldData.shouldSignal ? Redstone.SIGNAL_MIN : state.getSignal(level, pos, direction);
     }
 
     @Override
     protected int getSignal(final BlockState state, final BlockGetter level, final BlockPos pos, final Direction direction) {
-        if (this.shouldSignal && direction != Direction.DOWN) {
+        net.sparklypower.sparklypaper.threadedregions.GenericWorldData worldData = net.sparklypower.sparklypaper.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
+        if (worldData.shouldSignal && direction != Direction.DOWN) {
             int power = this.ownSignal(state, level, pos);
             if (power == 0) {
                 return Redstone.SIGNAL_MIN;
@@ -463,7 +466,8 @@ public class RedstoneWireBlock extends Block {
 
     @Override
     protected boolean isSignalSource(final BlockState state) {
-        return this.shouldSignal;
+        net.sparklypower.sparklypaper.threadedregions.GenericWorldData worldData = net.sparklypower.sparklypaper.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
+        return worldData.shouldSignal;
     }
 
     public static int getColorForPower(final int power) {
diff --git a/net/minecraft/world/level/block/SaplingBlock.java b/net/minecraft/world/level/block/SaplingBlock.java
index 5564a54f23aab158b9506325953ec70cab55c961..4a088fc85cf1171bd6ebdbdc4eba8c97affba22c 100644
--- a/net/minecraft/world/level/block/SaplingBlock.java
+++ b/net/minecraft/world/level/block/SaplingBlock.java
@@ -21,7 +21,7 @@ public class SaplingBlock extends VegetationBlock implements BonemealableBlock {
     public static final int TICK_CHANCE_FOR_SAPLING_GROWTH = 7;
     private static final VoxelShape SHAPE = Block.column(12.0, 0.0, 12.0);
     protected final TreeGrower treeGrower;
-    public static org.bukkit.@org.jspecify.annotations.Nullable TreeType treeType; // CraftBukkit
+    public static final ThreadLocal<org.bukkit.TreeType> treeTypeRT = new ThreadLocal<>(); // SparklyPaper - parallel world ticking (from Folia)
 
     protected SaplingBlock(final TreeGrower treeGrower, final BlockBehaviour.Properties properties) {
         super(properties);
@@ -60,8 +60,8 @@ public class SaplingBlock extends VegetationBlock implements BonemealableBlock {
                     capturedBlockStates = new java.util.ArrayList<>(level.capturedBlockStates.values());
                     level.capturedBlockStates.clear();
 
-                    treeType = SaplingBlock.treeType;
-                    SaplingBlock.treeType = null;
+                    treeType = SaplingBlock.treeTypeRT.get(); // SparklyPaper - parallel world ticking
+                    SaplingBlock.treeTypeRT.set(null);
                 }
                 if (!capturedBlockStates.isEmpty()) {
                     org.bukkit.Location location = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level);
diff --git a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
index c2f8479eef826efca3ed2993973b2cfc2c51ed00..e4a9601ba09284fad2225447a5eb3a971dc99d81 100644
--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -72,6 +72,12 @@ public abstract class BaseContainerBlockEntity extends BlockEntity implements Co
     protected abstract Component getDefaultName();
 
     public boolean canOpen(final Player player) {
+        // SparklyPaper - parallel world ticking (see: PARALLEL_NOTES.md - Opening an inventory after a world switch)
+        if (player instanceof net.minecraft.server.level.ServerPlayer serverPlayer && getLevel() != serverPlayer.level()) {
+            net.minecraft.server.MinecraftServer.LOGGER.warn("Player " + serverPlayer.getScoreboardName() + " (" + serverPlayer.getStringUUID() + ") attempted to open a BlockEntity @ " + this.getLevel().getWorld().getName() + " " + this.getBlockPos().getX() + ", " + this.getBlockPos().getY() + ", " + this.getBlockPos().getZ() + " while they were in a different world " + serverPlayer.level().getWorld().getName() + " than the block themselves!");
+            return false;
+        }
+        // SparklyPaper end
         return org.bukkit.craftbukkit.event.CraftEventFactory.callBlockLockCheckEvent(this, this.lockKey, this.getDisplayName(), player); // Paper - Call BlockLockCheckEvent
     }
 
diff --git a/net/minecraft/world/level/block/entity/BlockEntity.java b/net/minecraft/world/level/block/entity/BlockEntity.java
index ddda24a55201de2136fc6dd2cb863d609925d0ca..cc9a658ca0d0ed745333c44ebc09579d79208635 100644
--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -40,6 +40,7 @@ import org.slf4j.Logger;
 
 public abstract class BlockEntity implements DebugValueSource, TypedInstance<BlockEntityType<?>> {
     static boolean ignoreBlockEntityUpdates; // Paper - Perf: Optimize Hoppers
+    static final ThreadLocal<Boolean> ignoreBlockEntityUpdatesTL = ThreadLocal.withInitial(() -> Boolean.FALSE); // SparklyPaper - parallel world ticking
     // CraftBukkit start - data containers
     private static final org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry();
     public final org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer persistentDataContainer;
@@ -210,7 +211,7 @@ public abstract class BlockEntity implements DebugValueSource, TypedInstance<Blo
 
     public void setChanged() {
         if (this.level != null) {
-            if (ignoreBlockEntityUpdates) return; // Paper - Perf: Optimize Hoppers
+            if (ignoreBlockEntityUpdatesTL.get()) return; // Paper - Perf: Optimize Hoppers // SparklyPaper - parallel world ticking
             setChanged(this.level, this.worldPosition, this.blockState);
         }
     }
diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 950436f075c5ee01e15e71614c6cdc4bc5652fcc..91c67242cfba7b9e4c2f553982654987b4736b09 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -223,7 +223,8 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     private static boolean skipPushModeEventFire;
 
     private static boolean hopperPush(final Level level, final Container destination, final Direction direction, final HopperBlockEntity hopper) {
-        skipPushModeEventFire = skipHopperEvents;
+        final net.sparklypower.sparklypaper.threadedregions.GenericWorldData worldData = net.sparklypower.sparklypaper.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
+        worldData.skipPushModeEventFire = worldData.skipHopperEvents;
         boolean foundItem = false;
         for (int i = 0; i < hopper.getContainerSize(); ++i) {
             final ItemStack item = hopper.getItem(i);
@@ -246,7 +247,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
                 // We only need to fire the event once to give protection plugins a chance to cancel this event
                 // Because nothing uses getItem, every event call should end up the same result.
-                if (!skipPushModeEventFire) {
+                if (!worldData.skipPushModeEventFire) {
                     movedItem = callPushMoveEvent(destination, movedItem, hopper);
                     if (movedItem == null) { // cancelled
                         origItemStack.setCount(originalItemCount);
@@ -286,7 +287,8 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         container.setChanged(); // original logic always marks source inv as changed even if no move happens.
         movedItem.setCount(movedItemCount);
 
-        if (!skipPullModeEventFire) {
+        final net.sparklypower.sparklypaper.threadedregions.GenericWorldData worldData = net.sparklypower.sparklypaper.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
+        if (!worldData.skipPullModeEventFire) {
             movedItem = callPullMoveEvent(hopper, container, movedItem);
             if (movedItem == null) { // cancelled
                 origItemStack.setCount(originalItemCount);
@@ -306,9 +308,9 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 origItemStack.setCount(originalItemCount - movedItemCount + remainingItemCount);
             }
 
-            ignoreBlockEntityUpdates = true;
+            ignoreBlockEntityUpdatesTL.set(true); // SparklyPaper - parallel world ticking
             container.setItem(i, origItemStack);
-            ignoreBlockEntityUpdates = false;
+            ignoreBlockEntityUpdatesTL.set(false); // SparklyPaper - parallel world ticking
             container.setChanged();
             return true;
         }
@@ -332,7 +334,8 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         );
         final boolean result = event.callEvent();
         if (!event.calledGetItem && !event.calledSetItem) {
-            skipPushModeEventFire = true;
+            final net.sparklypower.sparklypaper.threadedregions.GenericWorldData worldData = net.sparklypower.sparklypaper.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
+            worldData.skipPushModeEventFire = true;
         }
         if (!result) {
             applyCooldown(hopper);
@@ -348,6 +351,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
     @Nullable
     private static ItemStack callPullMoveEvent(final Hopper hopper, final Container container, final ItemStack itemstack) {
+        final net.sparklypower.sparklypaper.threadedregions.GenericWorldData worldData = net.sparklypower.sparklypaper.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
         final org.bukkit.inventory.Inventory sourceInventory = getInventory(container);
         final org.bukkit.inventory.Inventory destination = getInventory(hopper);
 
@@ -355,7 +359,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         final io.papermc.paper.event.inventory.PaperInventoryMoveItemEvent event = new io.papermc.paper.event.inventory.PaperInventoryMoveItemEvent(sourceInventory, org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitMirror(itemstack), destination, false);
         final boolean result = event.callEvent();
         if (!event.calledGetItem && !event.calledSetItem) {
-            skipPullModeEventFire = true;
+            worldData.skipPullModeEventFire = true;
         }
         if (!result) {
             applyCooldown(hopper);
@@ -523,8 +527,9 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         BlockState blockState = level.getBlockState(blockPos);
         Container container = getSourceContainer(level, hopper, blockPos, blockState);
         if (container != null) {
+            final net.sparklypower.sparklypaper.threadedregions.GenericWorldData worldData = net.sparklypower.sparklypaper.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
             Direction direction = Direction.DOWN;
-            skipPullModeEventFire = skipHopperEvents; // Paper - Perf: Optimize Hoppers
+            worldData.skipPullModeEventFire = worldData.skipHopperEvents; // Paper - Perf: Optimize Hoppers
 
             for (int slot : getSlots(container, direction)) {
                 if (tryTakeInItemFromSlot(hopper, container, slot, direction, level)) { // Spigot
@@ -628,9 +633,9 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                     itemStack = itemStack.split(container.getMaxStackSize());
                 }
                 // Spigot end
-                ignoreBlockEntityUpdates = true; // Paper - Perf: Optimize Hoppers
+                ignoreBlockEntityUpdatesTL.set(true); // Paper - Perf: Optimize Hoppers // SparklyPaper - parallel world ticking
                 container.setItem(slot, itemStack);
-                ignoreBlockEntityUpdates = false; // Paper - Perf: Optimize Hoppers
+                ignoreBlockEntityUpdatesTL.set(false); // Paper - Perf: Optimize Hoppers // SparklyPaper - parallel world ticking
                 itemStack = leftover; // Paper - Make hoppers respect inventory max stack size
                 success = true;
             } else if (canMergeItems(current, itemStack)) {
diff --git a/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java b/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
index a1b0168077b83575f16dd377da2e1ba520d3bd19..7481cc1a8b1300c19c69bde46c9aa09acb45859a 100644
--- a/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -43,9 +43,9 @@ public class SculkCatalystBlockEntity extends BlockEntity implements GameEventLi
     // Paper end - Fix NPE in SculkBloomEvent world access
 
     public static void serverTick(final Level level, final BlockPos pos, final BlockState state, final SculkCatalystBlockEntity entity) {
-        org.bukkit.craftbukkit.event.CraftEventFactory.sourceBlockOverride = pos; // CraftBukkit - SPIGOT-7068: Add source block override, not the most elegant way but better than passing down a BlockPos up to five methods deep.
+        org.bukkit.craftbukkit.event.CraftEventFactory.sourceBlockOverrideRT.set(pos); // SparklyPaper - parallel world ticking // CraftBukkit - SPIGOT-7068: Add source block override, not the most elegant way but better than passing down a BlockPos up to five methods deep.
         entity.catalystListener.getSculkSpreader().updateCursors(level, pos, level.getRandom(), true);
-        org.bukkit.craftbukkit.event.CraftEventFactory.sourceBlockOverride = null; // CraftBukkit
+        org.bukkit.craftbukkit.event.CraftEventFactory.sourceBlockOverrideRT.set(null); // SparklyPaper - parallel world ticking // CraftBukkit
     }
 
     @Override
diff --git a/net/minecraft/world/level/block/grower/TreeGrower.java b/net/minecraft/world/level/block/grower/TreeGrower.java
index 95ac67eeee1c5c12fd709fe44288d6de9a28b3a1..3c3ef88066b72e4e1a530274c990111d57c45c5c 100644
--- a/net/minecraft/world/level/block/grower/TreeGrower.java
+++ b/net/minecraft/world/level/block/grower/TreeGrower.java
@@ -234,57 +234,59 @@ public final class TreeGrower {
 
     // CraftBukkit start
     private void setTreeType(Holder<Feature> feature) {
+        org.bukkit.TreeType treeType; // SparklyPaper - parallel world ticking
         if (feature.is(TreeFeatures.OAK) || feature.is(TreeFeatures.OAK_BEES_005)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TREE;
+            treeType = org.bukkit.TreeType.TREE; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.HUGE_RED_MUSHROOM)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.RED_MUSHROOM;
+            treeType = org.bukkit.TreeType.RED_MUSHROOM; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.HUGE_BROWN_MUSHROOM)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BROWN_MUSHROOM;
+            treeType = org.bukkit.TreeType.BROWN_MUSHROOM; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.JUNGLE_TREE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.COCOA_TREE;
+            treeType = org.bukkit.TreeType.COCOA_TREE; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.JUNGLE_TREE_NO_VINE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SMALL_JUNGLE;
+            treeType = org.bukkit.TreeType.SMALL_JUNGLE; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.PINE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_REDWOOD;
+            treeType = org.bukkit.TreeType.TALL_REDWOOD; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.SPRUCE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.REDWOOD;
+            treeType = org.bukkit.TreeType.REDWOOD; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.ACACIA)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.ACACIA;
+            treeType = org.bukkit.TreeType.ACACIA; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.BIRCH) || feature.is(TreeFeatures.BIRCH_BEES_005)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIRCH;
+            treeType = org.bukkit.TreeType.BIRCH; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.SUPER_BIRCH_BEES_0002)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_BIRCH;
+            treeType = org.bukkit.TreeType.TALL_BIRCH; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.SWAMP_OAK)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SWAMP;
+            treeType = org.bukkit.TreeType.SWAMP; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.FANCY_OAK) || feature.is(TreeFeatures.FANCY_OAK_BEES_005)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIG_TREE;
+            treeType = org.bukkit.TreeType.BIG_TREE; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.JUNGLE_BUSH)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE_BUSH;
+            treeType = org.bukkit.TreeType.JUNGLE_BUSH; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.DARK_OAK)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.DARK_OAK;
+            treeType = org.bukkit.TreeType.DARK_OAK; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.MEGA_SPRUCE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+            treeType = org.bukkit.TreeType.MEGA_REDWOOD; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.MEGA_PINE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_PINE;
+            treeType = org.bukkit.TreeType.MEGA_PINE; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.MEGA_JUNGLE_TREE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE;
+            treeType = org.bukkit.TreeType.JUNGLE; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.AZALEA_TREE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.AZALEA;
+            treeType = org.bukkit.TreeType.AZALEA; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.MANGROVE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MANGROVE;
+            treeType = org.bukkit.TreeType.MANGROVE; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.TALL_MANGROVE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_MANGROVE;
+            treeType = org.bukkit.TreeType.TALL_MANGROVE; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.CHERRY) || feature.is(TreeFeatures.CHERRY_BEES_005)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.CHERRY;
+            treeType = org.bukkit.TreeType.CHERRY; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.PALE_OAK) || feature.is(TreeFeatures.PALE_OAK_BONEMEAL)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.PALE_OAK;
+            treeType = org.bukkit.TreeType.PALE_OAK; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.PALE_OAK_CREAKING)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.PALE_OAK_CREAKING;
+            treeType = org.bukkit.TreeType.PALE_OAK_CREAKING; // SparklyPaper - parallel world ticking
         } else if (feature.is(TreeFeatures.RED_POPLAR) || feature.is(TreeFeatures.ORANGE_POPLAR) || feature.is(TreeFeatures.YELLOW_POPLAR)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.POPLAR;
+            treeType = org.bukkit.TreeType.POPLAR; // SparklyPaper - parallel world ticking
         }  else {
             throw new IllegalArgumentException("Unknown tree generator " + feature);
         }
+        net.minecraft.world.level.block.SaplingBlock.treeTypeRT.set(treeType); // SparklyPaper - parallel world ticking
     }
     // CraftBukkit end
 }
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index af4106497a69f9b311970197016c6157475a7d85..283eabd4e2ad98439d329ef334ab63ed10d46c82 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -364,6 +364,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
     @Override
     public @Nullable BlockState setBlockState(final BlockPos pos, final BlockState state, final @Block.UpdateFlags int flags) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, pos, "Updating block asynchronously"); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         int y = pos.getY();
         LevelChunkSection section = this.getSection(this.getSectionIndex(y));
         boolean wasEmpty = section.hasOnlyAir();
diff --git a/net/minecraft/world/level/entity/EntityTickList.java b/net/minecraft/world/level/entity/EntityTickList.java
index a30409eddfaded878a3ac54bb526fd2e4b887c57..d7b959d91ad7d5e018f356345e007533dc43956b 100644
--- a/net/minecraft/world/level/entity/EntityTickList.java
+++ b/net/minecraft/world/level/entity/EntityTickList.java
@@ -10,17 +10,26 @@ import org.jspecify.annotations.Nullable;
 
 public class EntityTickList {
     private final ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<net.minecraft.world.entity.Entity> entities = new ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<>(); // Paper - rewrite chunk system
+    // SparklyPaper start - parallel world ticking
+    // Used to track async entity additions/removals/loops
+    private final net.minecraft.server.level.ServerLevel serverLevel;
+    public EntityTickList(net.minecraft.server.level.ServerLevel serverLevel) {
+        this.serverLevel = serverLevel;
+    }
+    // SparklyPaper end
 
     private void ensureActiveIsNotIterated() {
         // Paper - rewrite chunk system
     }
 
     public void add(final Entity entity) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(entity, "Asynchronous entity ticklist addition"); // Paper // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         this.ensureActiveIsNotIterated();
         this.entities.add(entity); // Paper - rewrite chunk system
     }
 
     public void remove(final Entity entity) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(entity, "Asynchronous entity ticklist removal"); // Paper // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         this.ensureActiveIsNotIterated();
         this.entities.remove(entity); // Paper - rewrite chunk system
     }
@@ -30,6 +39,7 @@ public class EntityTickList {
     }
 
     public void forEach(final Consumer<Entity> output) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(serverLevel, "Asynchronous entity ticklist iteration"); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
         // Paper start - rewrite chunk system
         // To ensure nothing weird happens with dimension travelling, do not iterate over new entries...
         // (by dfl iterator() is configured to not iterate over new entries)
diff --git a/net/minecraft/world/level/saveddata/maps/MapIndex.java b/net/minecraft/world/level/saveddata/maps/MapIndex.java
index 6b3ce2ded53a7ea46cf7f0cb49043a022f6d9ee7..8044377482d8adbef28bb6239ad73061304678ed 100644
--- a/net/minecraft/world/level/saveddata/maps/MapIndex.java
+++ b/net/minecraft/world/level/saveddata/maps/MapIndex.java
@@ -26,8 +26,10 @@ public class MapIndex extends SavedData {
     }
 
     public MapId getNextMapId() {
+        synchronized (TYPE) { // SparklyPaper start - parallel world ticking
         MapId id = new MapId(++this.lastMapId);
         this.setDirty();
         return id;
+        } // SparklyPaper end
     }
 }
