From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Fri, 6 Feb 2026 16:04:58 -0300
Subject: [PATCH] Add WorldTickStartEvent and WorldTickEndEvent


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 227e91de018cfd2889d5dea2f2b97faa2e12df9d..dccc58e478ae3933eab54f2323a9eb654540e7ab 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1831,11 +1831,15 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                         ca.spottedleaf.moonrise.common.util.TickThread.ServerLevelTickThread currentThread = (ca.spottedleaf.moonrise.common.util.TickThread.ServerLevelTickThread) Thread.currentThread();
                         currentThread.currentlyTickingServerLevel = level;
 
+                        new net.sparklypower.sparklypaper.event.world.WorldTickStartEvent(level.getWorld(), this.tickCount).callEvent();
+
                         long i = Util.getNanos(); // SparklyPaper - track world's MSPT
                         level.tick(haveTime);
                         // SparklyPaper start - track world's MSPT
                         long j = Util.getNanos() - i;
 
+                        new net.sparklypower.sparklypaper.event.world.WorldTickEndEvent(level.getWorld(), this.tickCount, j).callEvent();
+                        
                         // These are from the "tickServer" function
                         level.tickTimes5s.add(this.tickCount, j);
                         level.tickTimes10s.add(this.tickCount, j);
